src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/CannotRemoveBaseTypeOrInterface.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Microsoft.DotNet.ApiCompatibility.Extensions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    public class CannotRemoveBaseTypeOrInterface : Rule
+    {
+        public override void Initialize(RuleRunnerContext context)
+        {
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left == null || right == null)
+                return;
+
+            if (left.TypeKind != TypeKind.Interface && right.TypeKind != TypeKind.Interface)
+            {
+                // if left and right are not interfaces check base types
+                ValidateBaseTypeNotRemoved(left, right, leftName, rightName, differences);
+
+                if (Settings.StrictMode)
+                    ValidateBaseTypeNotRemoved(right, left, rightName, leftName, differences);
+            }
+
+            ValidateInterfaceNotRemoved(left, right, leftName, rightName, differences);
+
+            if (Settings.StrictMode)
+                ValidateInterfaceNotRemoved(right, left, rightName, leftName, differences);
+        }
+        private void ValidateBaseTypeNotRemoved(ITypeSymbol left, ITypeSymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            // Base type order matters in the hierarchy chain so therefore use a list.
+            List<ITypeSymbol> rightBaseTypes = right.GetAllBaseTypes().ToList();
+
+            int lastTypeIndex = 0;
+            foreach (ITypeSymbol leftBaseType in left.GetAllBaseTypes())
+            {
+                lastTypeIndex = rightBaseTypes.FindIndex(lastTypeIndex, r => Settings.SymbolComparer.Equals(leftBaseType, r));
```

Comment:
can you explain why the start point for search is lastTypeIndex ? how does it ensures the hierarchy

Reply:
Sure. The base class hierarchy order can not change, as in order to change the order you would be dropping a base class from another class at some point and that is a breaking change, so we are guaranteed that we can check the base classes from bottom to top on the tree skipping the indexes we already checked.

i.e if you have:

Left:
```cs
public class Foo : FirstBase { }
public class FirstBase : SecondBase { }
public class SecondBase : ThirdBase { }
public class ThirdBase { }
```

if I were to introduce a new base class I would have to honor order and push the previous base class up in the hierarchy:
Right: 
```cs
public class Foo : FirstBase { }
public class FirstBase : NewSecondBase { }
public class NewSecondBase : SecondBase { }
public class SecondBase : ThirdBase { }
public class ThirdBase { }
```

That would be flattened as
Right: `[FirstBase, NewSecondBase, SecondBase, ThirdBase]`
Left: `[FirstBase, SecondBase, ThirdBase]`.

So that way you when we are looking for `ThirdBase` in `Right`, `lastTypeIndex=2`, so we skip the base types we don't care and that way it is faster to find all the base types.

Doing something like the following is a breaking change:
```cs
public class First : FirstBase { }
public class FirstBase : SecondBase { }
public class SecondBase { }
```
changed to:
```cs
public class First : FirstBase, SecondBase { }
public class FirstBase { }
public class SecondBase { }
```
Is breaking as `FirstBase` no longer has `SecondBase` as it's base class.